fvm/src/engine.rs

```diff
@@ -619,6 +638,112 @@ impl Engine {
     }
 }
 
+pub struct WasmInstance {
+    instance: wasmtime::Instance,
+    pool: Arc<InstancePool>,
+}
+
+impl Deref for WasmInstance {
+    type Target = wasmtime::Instance;
+
+    fn deref(&self) -> &Self::Target {
+        &self.instance
+    }
+}
+
+impl Drop for WasmInstance {
+    fn drop(&mut self) {
+        self.pool.put()
+    }
+}
+
+struct InstancePoolInner {
+    // available resources
+    avail: u32,
+    // resource reservation limit; if avail < rsvp, the resource pool will boost one thread
+    // to receive as many resources as it needs, while avail > 0.
+    // if avail reaches 0, the resource reservation will fail.
+    rsvp: u32,
+    // thread id of currently boosted thread, if any
+    boost: Option<thread::ThreadId>,
+    // active boosts for currently boosted threads
+    boosting: u32,
+}
+
+struct InstancePool {
+    mx: Mutex<InstancePoolInner>,
+    cv: Condvar,
+}
+
+impl InstancePool {
+    fn new(avail: u32, rsvp: u32) -> InstancePool {
+        InstancePool {
+            mx: Mutex::new(InstancePoolInner {
+                avail,
+                rsvp,
+                boost: None,
+                boosting: 0,
+            }),
+            cv: Condvar::new(),
+        }
+    }
+
+    fn get(&self) -> Result<(), Abort> {
```

Comment:
This logic isn't _quite_ optimal. We don't need to reserve an additional `rsvp` for the boosted executor, we need to make sure that the boosted executor has `rsvp` available _total_, including the instances it already has.

Right now, we could end up reserving up to `2*rsvp` for said executor.

Reply:
I know, but we would have to track how much eaxh executor is using then, which complicates things.

I'd rather have some slack that is actually covered by the excutor's own logic (it does check the stack depth) than overcomplicate in this component.

Reply:
Maybe we are fine with tracking however, it is just a map of executor id to resource usage.

Probably not thst complicated.

If you feel strongly about it, I can add tracking.